### DISCUSSION TITLE
LR1110 Gnss scan status

### DISCUSSION THREAD

**Discussion by tve** [jgromes/RadioLib#1211] [General]
First le me say thank you for having started to implement the Gnss scanning for the LR1110!!! I got a WM1110 Tracker dev board and tried to use the Seed stuff, which was a total no-go 'cause they're stuck on the 3.x firmware, which doesn't support autonomous almanac updates or calculating a 2D fix. I spent a day trying to upgrade the software but it's just such a nightmare... Eventually it dawned on me that I might want to check RadioLib...

Anyway, I'm using master and am wondering whether there is any further progress in some fork or other repo?

If not, that's OK, I'll just plug away at implementing the stuff that I find missing. I've adapted the WiFi scan example and I now see a GNSS scan result in the debug output but so far it's not seeing any SV after 63 secs  (I'm indoors with a "repeater" antenna and my phone does see the expected number of SVs). I'll start by reading the LR1110 docs to understand where I need to start...

**Comment by tve** [comment]
Aha! Found this in the Seeed WM1110_Geolocation repo:
```
    lbmWm1110.attachGnssPrescan([](void* context){ digitalWrite(PIN_GNSS_LNA, HIGH); });
    lbmWm1110.attachGnssPostscan([](void* context){ digitalWrite(PIN_GNSS_LNA, LOW); });
```
Updated the rf switch table:
```
// set RF switch configuration for Wio Tracker with Wio WM1110 Module
// Wio WM1110 uses DIO5 and DIO6 for RF switching
// Wio Tracker 1110 dev board adds LNA for GNSS antenna controlled by PIN_GNSS_LNA
// NOTE: other boards may be different!
static const uint32_t rfswitch_dio_pins[] = { 
  RADIOLIB_LR11X0_DIO5, RADIOLIB_LR11X0_DIO6,
  PIN_GNSS_LNA, RADIOLIB_NC, RADIOLIB_NC
};

static const Module::RfSwitchMode_t rfswitch_table[] = {
  // mode                  DIO5  DIO6 GNSS
  { LR11x0::MODE_STBY,   { LOW,  LOW, LOW } },
  { LR11x0::MODE_RX,     { HIGH, LOW, LOW } },
  { LR11x0::MODE_TX,     { HIGH, HIGH,LOW } },
  { LR11x0::MODE_TX_HP,  { LOW,  HIGH,LOW } },
  { LR11x0::MODE_TX_HF,  { LOW,  LOW, LOW } },
  { LR11x0::MODE_GNSS,   { LOW,  LOW, HIGH } },
  { LR11x0::MODE_WIFI,   { LOW,  LOW, LOW } },
  END_OF_MODE_TABLE,
};
```
But the GNSS scan code doesn't set the rf switch table so have to add that... And the LR11x0::setRfSwitchTable doesn't seems to allow GPIO pins to be used. Hmmm... Gonna turn it on globally for now...

---

With the LNA turned on I'm finally seeing some SVs:
```
RLB_DBG: Detected 8 SVs:
RLB_DBG:   SV 0: C/N0 43 dB, Doppler 3476 Hz
RLB_DBG:   SV 1: C/N0 42 dB, Doppler -860 Hz
RLB_DBG:   SV 2: C/N0 40 dB, Doppler 709 Hz
RLB_DBG:   SV 3: C/N0 40 dB, Doppler 3706 Hz
RLB_DBG:   SV 4: C/N0 38 dB, Doppler 4368 Hz
RLB_DBG:   SV 5: C/N0 36 dB, Doppler 3502 Hz
RLB_DBG:   SV 6: C/N0 35 dB, Doppler 548 Hz
RLB_DBG:   SV 7: C/N0 34 dB, Doppler 4516 Hz
```
But I mostly get a "demod status: -17" which the manual says is "20 ms real time failure". So far I have not found what that means... Other times I get 'The satellite selected for demodulation lost' (sic), or 'No bit change found during advanced scan', or 'Too many parity check found' (sic).
At one point I did see 'Week Number found (Time Of Week as well)' but nothing more came of that.

I'm wondering whether there's something I'm missing that is required to cold-start the lr1110...

---

Hmm, GnssReadTime says "No 32kHz clock available". Gotta look at schematics to see what I need to do to enable that.

---

My WIP code is in the lr1110-gnss branch of my fork https://github.com/tve/RadioLib/tree/lr1110-gnss

**Reply by jgromes** [reply]
GNSS scanning for the LR1110 is work in progress. At some point, I accidentaly commited the files into master, which I will fix before the next release and move this stuff into its own branch.

I was able to get time from the satellites, but got stuck trying to update the almanac, which I suspect is required to get the internal solver to work. The main issue I'm facing is the documentation blackout for this functionality, there is very little in the user manual and nearly nothing in the Semtech repos. So I've mostly been trying to guess how different parts of teh scanning work.

You have correctly identified that there is an on-board LNA for the WM1110, and yes, RadioLib does not handle that yet.

> GnssReadTime says "No 32kHz clock available"

Been there already, you have to configure the LF clock to use the 32.768k crystal using `LR11x0::configLfClock`.

**Reply by tve** [reply]
I'm glad you committed by mistake! Otherwise I wouldn't have picked it up and been stuck with Semtech's LBM!

> Been there already, you have to configure the LF clock to use the 32.768k crystal using R11x0::configLfClock.

I saw that and did that but it's not helping. Maybe I'm doing something wrong. Am going through the various init steps again.

Is lora-developers.semtech.com shut down? I seem to remember that they discontinued something.. I guess too many inconvenient questions...

(Dunno whether you saw that I added a link to my fork a few minutes before you responded)

**Reply by jgromes** [reply]
> I saw that and did that but it's not helping. Maybe I'm doing something wrong

You're not, there's a mistake in `configLfClock`, it sends the wrong command, `RADIOLIB_LR11X0_CMD_SET_REG_MODE` instead of `RADIOLIB_LR11X0_CMD_CONFIG_LF_LOCK`. I pushed a fix for that (amoung a few other command fixes), so that nobody gets stuck on known issues.

> Is lora-developers.semtech.com shut down? I seem to remember that they discontinued something.. I guess too many inconvenient questions...

I think it's been down for quite some time now, more than a year. I seem to recall it was getting heavily spammed, maybe even hacked since some of the spamming accounts were mods IIRC. So I guess it was easiest just to shut it down and use GitHub repos instead.

**Reply by tve** [reply]
Thanks for the fix, it would have taken me a long time to figure this out!
I'm still stuck. If I call configLfClock(5) it succeeds but then all scans time out on the IRQ wait. If I use configLfClock(1) the scans work, but the demod isn't happy. I have 10 sats with C/N0 > 35 yet keep getting negative demod statuses, primarily -17 ("20ms real time failure").
Stumped.

Example:
```

[LR1110] ***** Starting GNSS scan ... 
RLB_DBG: GNSS scan start
RLB_DBG: Starting GNSS scan with effort=1 resMask=3c nbsvMax=8
RLB_DBG: GNSS scan done in 71241 ms
RLB_DBG: Demod status '20 ms real time failure' (-17), info 00
RLB_DBG: Context status fwVersion 2, almanacCrc f9e87ffe, errCode 0, almUpdMask 3, freqSpace 0
RLB_DBG: GPS almanac status: 'No satellite to update' (0), SVs needing update: 0
RLB_DBG:     delay: 0; subframes: 0; SVs: 0,0; next subframe: 0
RLB_DBG: bdu almanac status: 'No satellite to update' (0), SVs needing update: 0
RLB_DBG:     delay: 0; subframes: 0; SVs: 0 0; next subframe: 0
[LR1110] Reading GNSS scan result ... 
RLB_DBG: Last scan mode 'Cold start mode (Time and Assisted Position unknown)' (4)
RLB_DBG: Detected 9 SVs:
RLB_DBG:   SV 0: C/N0 42 dB, Doppler -1665 Hz
RLB_DBG:   SV 1: C/N0 41 dB, Doppler 3928 Hz
RLB_DBG:   SV 2: C/N0 41 dB, Doppler 1554 Hz
RLB_DBG:   SV 3: C/N0 37 dB, Doppler 1934 Hz
RLB_DBG:   SV 4: C/N0 37 dB, Doppler 4382 Hz
RLB_DBG:   SV 5: C/N0 36 dB, Doppler 2661 Hz
RLB_DBG:   SV 6: C/N0 36 dB, Doppler -1204 Hz
RLB_DBG:   SV 7: C/N0 35 dB, Doppler 3574 Hz
RLB_DBG:   SV 8: C/N0 35 dB, Doppler 1034 Hz
RLB_DBG: GPS time not available (No 32kHz clock)
RLB_DBG: lat = 89.95605, lon = 0.00000
```

**Reply by tve** [reply]
I left the app running and scanning all night and it doesn't make progress. I always get a demod error -17, -20, or -7, despite consistently having 6+ satellites with C/N0 above 35dB. Something is not right... ~~Not sure what to try next.~~

I found out that after calling gnssScan I get a 0x40 error status from getErrors, that's "LF_OSC_START_ERR". When do you call ConfigLfOsc and what value do you use? Is there something else that's perhaps also needed?

I guess the hanging of gnssScan when I use ConfigLfClock(5) is consistent with the error: it doesn't start so busy remains active. I thought this would happen when calling ConfigLfClock but evidently it happens when calling GnssScan.

Apparently there must be a silicon bug 'cause their code says that if the LF OSC doesn't start up one has to reset the chip "the specification says". So sometimes it may not have worked for me due to that? Ughh.

**Reply by jgromes** [reply]
> I always get a demod error -17, -20, or -7, despite consistently having 6+ satellites with C/N0 above 35dB

That sounds like about the same point where I got stuck.

> When do you call ConfigLfOsc and what value do you use?

I created a new `beginGNSS` method to handle that - see https://github.com/jgromes/RadioLib/tree/lr11x0-gnss for details. Though I did try out multiple different values without much progress.

> I guess the hanging of gnssScan when I use ConfigLfClock(5) is consistent with the error: it doesn't start so busy remains active

That seems logical - but the question remains, why does LF OSC consistently fail to start up?

**Reply by tve** [reply]
Thanks for the update. I had thought you had gotten further. Good news/bad news I guess...
I had missed the new beginGnss, will use that.

To recap what I found:
- the LBM_WM1110 code uses ConfigLfClock(5): https://github.com/Seeed-Studio/LBM_WM1110/blob/ca8f39f81beba5465691567df60325355ba1d691/src/mw/geolocation_middleware/common/mw_common.c#L96
- if I use ConfigLfClock(5) -> succeeds, busy=0 after the call, no errors, but GnssScan IRQ times out with busy=1 and all one can do is reset the radio
- ConfigLfClock(1) -> succeeds, no errors afterwards, GnssScan sometimes(?) errors with 0x40 error -> LF_XTAL_START_ERR
- the LBM_WM1110 code checks errors right after calling ConfigLfClock, which in my testing doesn't result in an error even though GnssScan later hangs/errors

Update:
- I got curious about the 32kHz xtal on the tracker board... the WM1110 module block diagrams show two 32kHz xtals, so why another one on the board
- I decapped the module, turns out there's only one 32kHz xtal in there! for the LR1110, the one on the board is for the nRf

Update2:
One issue I've noticed is that DIO11 can be used both for xtal or as IRQ2 output. It's not entirely clear to me how one ensures it's available for xtal use and not IRQ2. In the current code, setDioIrqParams(x) sets flags for both IRQ1 and IRQ2 so I presume that tells the chip to use IRQ2/DIO11 for IRQ and not for xtal. But I've changed that to set IRQ2=0 and so far it doesn't seem to do it, but it also seems that the GNSS-DONE interrupt doesn't trigger DIO9... Grrrr!

**Reply by tve** [reply]
Well finally some success, but not without issues:
```
RLB_DBG: GNSS scan done in 67631 ms
RLB_DBG: Demod status 'Week Number found (Time Of Week as well)' (4), info 7f
```
I believe the trick was to not set DioIrq for DIO11. But I have a problem, which is that the IRQ pin is not triggered. I printed the buffer sent for setDioIrqParams and it has bit 19 (and 28) set and I set the pimode to output just before starting the GNSS scan and yet it stays 0. I added a GetStatus command into the wait loop and check the bit there as a temporary hack and that got me to the above...

Update: the scope says that the IRQ pin goes high, so it's the nRf side that is not working. While probing it it seemed to change behavior. Gave it a little soldering iron treatment. Seems to work now. Weird, maybe something else I don't understand changed...
Anyway, almanac: here I come!

**Reply by tve** [reply]
Yay! First Almanac update!
```
GPS almanac status: 'At least 1 satellite must be updated' (1), SVs needing update: 32
    delay: 10768ms; subframes: 2; SVs: 31,9; next subframe: 4
Updating almanac in ~10768ms
GPS almanac update done in 12828 ms
Almanac update done
GPS almanac status: 'At least 1 satellite must be updated' (1), SVs needing update: 31
    delay: 17375ms; subframes: 2; SVs: 32,10; next subframe: 4
```
Note that the "SVs needing update" went from 32 to 31 :-)
The timing for starting the update is relatively tight, I had added 500ms margin and it aborted and checking the almanac status again said to update the same SVs in ~140ms.
Time for some sleep...

**Reply by tve** [reply]
The full almanac got pulled in overnight and the approximate position is now calculated. It seems to bounced around about 6-7km from the actual position at the moment. I guess that's due to just using the almanac data and no ephemeris.
Next is getting the position data to lora-cloud to see how it handles things.
Also, I see the scans taking ~10 seconds so I'll have to see how to make it work with much shorter scans and how that comes out. (I haven't dug into how long the device is in RX during those 10 seconds.)

**Reply by jgromes** [reply]
Darn, seems like you got it working! Congratulations! Was there anything specific you had to change from the upstream to be able to update the almanac, or was it just timing? I did try waiting based on the timing information returned from `getGnssAlmanacStatus`, but that didn't seem to work.

**Reply by tve** [reply]
There were a bunch of little things to fix. I have to clean up my code a bit and will then push to my fork. I believe the main thing you were missing is that the Almanac update function has to be called at least 1.3 seconds before the actual deadline so it can lock onto the signal. (2.3s is the target in their code)

Now I'm wondering how I can erase the almanac data in order to be able to start over to test...

**Reply by jgromes** [reply]
It would be great to get those fixes in a PR once it's cleaned up ;)

> Now I'm wondering how I can erase the almanac data in order to be able to start over to test...

There's the `gnssAlmanacReadAddrSize` that returns the start and length of the almanac data, but I don't know whether there's an actual erase command. Fells like there should be one, at least in the bootloader, but there's nothing in the user manual. Alternatively, you could maybe perform the almanac update procedure with 0x00 or 0xFF data?

**Reply by tve** [reply]
I pushed my stuff with some minor clean-up. It's not a PR. Feel free to pull sections if you want to move forward. I'm still futzing around with stuff.

One item that is an issue in the lr1110 driver is the RfSwitchTable. It has moved into the chip but function like transmit and others still call the setRfSwitch method. In the case of the WM1110 tracker board that's actually needed 'cause the GNSS LNA is controlled by a gpio pin. But to actually make that function there's some more work required... (Right now I just set the LNA once for all in the app code)

I also noticed that there are several methods to retrieve the status from the device, that could be simplified into one. I had to fix one that didn't work at all.

Not done yet...

**Reply by tve** [reply]
I updated the PR adding a bash script to submit the NAV data to LoRaCloud and intersperse it into the demo script. I tried to keep this simple but it still takes a bunch of lines of code to make it work...

I'm seeing some oddities with the cloud stuff. There is some ambiguity about how `gnss_capture_time` is used and I have the feeling there are cases where omitting it produces a (good) result when omitting fails, have to test some more...

Unfortunately I have not been able to get any support from Semtech so far. Everything on their site leads to the contact form and I'm pretty sure they'll just keep ignoring me. If someone has a technical contact and could forward a handful of specific technical questions I'd very much appreciate the help!

Here's a sample output (The LoRaCloud section is added by the bash script):
```
[LR1110] Starting GNSS scan ... 

RLB_DBG: GNSS scan start 'medium effort'
RLB_DBG: Starting GNSS scan with effort=1 resMask=3c nbsvMax=8
RLB_DBG: Status: 5 CMD_OK, INTR, mode=1
RLB_DBG: IRQstatus=00080000 errors=0x0000
RLB_DBG:     IRQ bit 19 set
RLB_DBG: GNSS scan done in 10946 ms
RLB_DBG: Demod status 'Time Of Week found' (3), info 03
[LR1110] Reading GNSS scan result ...
RLB_DBG: Last scan mode 'Assisted mode (Time and Assisted Position known)' (3)
RLB_DBG: Detected 7 SVs:
RLB_DBG:   SV 0: SNR 12 dB, Doppler 300 Hz
RLB_DBG:   SV 1: SNR 12 dB, Doppler 3983 Hz
RLB_DBG:   SV 2: SNR 10 dB, Doppler 3959 Hz
RLB_DBG:   SV 3: SNR 5 dB, Doppler -931 Hz
RLB_DBG:   SV 4: SNR 5 dB, Doppler 2529 Hz
RLB_DBG:   SV 5: SNR 3 dB, Doppler 1815 Hz
RLB_DBG:   SV 6: SNR 3 dB, Doppler 2698 Hz
GPS time=1410390673 407205us accuracy=11397us unix=1726355455
NAV @1410390664: 631052c00000000c94e1b0000007dcc35d722e5565bbe78df146009
LoRaCloud:
  Position: 35.12345 -119.12345 alt: 560.39
  Real error: 26m
  Claimed accuracy: 110.5m
  GPS time: 1410390665.293538 -- delta=1s
  UTC time: 1726355447.293538 -- 2024-09-14T23:10:47+00:00
Demodulated position (using 7 SVs):
  One-shot: 35.999999 -119.970703 | accuracy:39[?] xtal:1046ppb | real err: 14721m
  Filtered: 35.584961 -119.707031 | accuracy:51[?] xtal:1035ppb | real err: 13937m
Last scan timing, power, and % of power:
  GPS: total 4.3s/55.8mJ | P1:45.5% Adv:54.5% | capture:38.7% cpu:51.2% sleep:10.0%
  Beidou: total 0.0s/0.0mJ | P1:nan% Adv:nan% | capture:nan% cpu:nan% sleep:nan%
  GPS demod: total 6.7s/56.2mJ | capture:46.0% cpu:8.2% sleep:45.8%
  TOTAL: 11.0s/112.0mJ | GPS:49.8% Beidou:0.0% Demod:50.1% Init:0.0%
GPS almanac status: 'At least 1 satellite must be updated' (1)
  SVs needing update: 1 (SVs 63 25)
  Update in 364.7s, first subframe 4, 2 subframes
BDU almanac status: 'No satellite to update' (0)
  SVs needing update: 0 (SVs 0 0)
  Update in 0.0s, first subframe 0, 0 subframes
```

**Reply by jgromes** [reply]
To be honest, my goal when starting with the LR11x0 GNSS scanning was to make it completely independent of the could component. I profoundly dislike the business model of "to be able to fully use the IC you paid money for, you have to register in our cloud service that probably won't be running in 3 years or less".

If it was optional, then sure, some users might like the convenience of CLOUD-based GEOlocation (funny as that terms is once you think about it). But not having the option to run my own server that would be able to provide almanac updates and decode the NAV data, that's just a big nope from my side. Hence me trying (and you succeeding ;) ) in getting it to run standalone. So I will strongly advocate in favor of keeping RadioLib free from all that interaction with the Semtech cloud.

It seems that there is still some ongoing development in the PR - please let me know once it's at a point where you'd like me to review and talk about how to merge it. We might have to think a bit how to best handle the public interface and the example(s).

**Reply by tve** [reply]
Hey, I totally sympathize with the subscription-free goal. At the moment I don't see an non-cloud option for evaluating the data coming out of the LR1110 (or a U-Blox gen 10 chip either). The only thing in the PR that has to do with the cloud service is the sample shell sript and it's there primarily for evaluation purposes. Now, to be honest, from what I've seen, if you don't use the cloud service I'm not sure why you would use the LR1110. The sample sketch and shell script let you specify the exact location of the device (assuming it's stationary) and it prints the location error. Using the cloud service you get the typical 8-20m error. Using the solver built into the chip you get anywhere from 8km to 30km error. The uses case where that amount of error is acceptable are not that many...

Anyway, at this stage I've moved on to sending the data from the script to a database and evaluating the performance of the LR1110 under various conditions, so not something that I will be adding to the PR. I'm still trying to understand how the ingredients to the GNSS work and have just added the list of open questions I have to the PR. Some of the open questions may affect what could be exposed in the API. But I expect my rate of change to slow down dramatically now.

**Reply by HeadBoffin** [reply]
@miguelcz123, this thread is all about GNSS on the LR1110. Communication between a module & an MCU is totally different and also well covered in the wiki. I've started an issue for you on this.

**Comment by tve** [comment]
I'm starting to accumulate some questions about the LR1110 operation for which I can't find answers... I've submitted a support request but have not heard back and may not hear back since I don't work for some big ass company... So far I have:

- is there any clean method for resetting the almanac data?
- in the data returned by GnssReadDopplerSolverRes what are the units for the accuracy value?
- in the data returned by GnssReadDopplerSolverRes what filtering is applied to the FilteredLongitude/Latitude?

---

WRT the filtering, after a cold-start I've noticed the following:
#### first fix:
```
Demodulated position (using 4 SVs):
  One-shot: 34.541016 -119.882812 | accuracy:54[?] xtal:1018ppb | real err: 7517m                    
  Filtered: 34.541016 -119.882812 | accuracy:54[?] xtal:1018ppb | real err: 7517m                    
```
(the real error is based on me having hard-coded the real GPS location and using the haversine formula)

####  second fix:
```
Demodulated position (using 4 SVs):          
  One-shot: 35.551758 -120.146484 | accuracy:95[?] xtal:982ppb | real err: 120779m 
  Filtered: 34.584961 -119.882812 | accuracy:56[?] xtal:1017ppb | real err: 11208m 
```

What I find significant is:
- the first filtered coords are the same as the one-shot values
- the second FilteredLongitude is the same as the first fix's while the FilteredLatitude is not and it's not an average either

**Comment by klym-tatarnikov** [comment]
Hi Team, 
Has anyone managed to achieve reasonable results?
Tried to use Seeed WM1110 tracker with latest Radiolib and latest firmware, some results, conclusions:

1. Both almanacs for GPS and Beidou have been updated (from sats).
2. GNSS time is 15s ahead from the real one (?). 
3. 'Accuracy' field UOM is not clear. 
4. Internally calculated location is +-100km even with 8-10 visible sats. 
5. Values received from internal solver seems to be truncated or lost accuracy during math conversions, see map below. 
6. T1000-A/B/E devices also based on LR1110 but still using separate GNSS chip (maybe they know something)
7. Semtech Lora Cloud has been decommissioned  on July 2025, so cloud solving is now supported byt thirdparty vendors. 
8. Even if you  want to use cloud-assited GNSS but with your own cloud - it's not possible as both Almanac structure and NAV3 messages are not documented and possible encrypted.

GNSS autotonous results from Seeed WIO Tracker (almost clear sky, 3 meters away from the building). Actual position is 10km away from closest point in this map,  
<img width="2093" height="712" alt="Screenshot 2025-09-07 110255" src="https://github.com/user-attachments/assets/8716edb6-a49f-4814-8528-73f589aead17" />

**Reply by jgromes** [reply]
@klym-tatarnikov unfortunately I think the current state is as good as it gets. The on-chip solver in the LR11x0 is pretty much a gimmick, not very useful for real-world tracking (unless you are fine with tracking precision of "roughly knowing which country you are in").

> Values received from internal solver seems to be truncated or lost accuracy during math conversions, see map below.

Indeed, that is the accuracy we get from the LR11x0, as seen below:

https://github.com/jgromes/RadioLib/blob/6c38520393e4714d2ba560245c2aca8ac6ed0487/src/modules/LR11x0/LR11x0_gnss.cpp#L636-L643

> Semtech Lora Cloud has been decommissioned on July 2025, so cloud solving is now supported byt thirdparty vendors.

Well would you look at that, it's almost as if someone called it a year ago in this very thread ... 🙃 

> Even if you want to use cloud-assited GNSS but with your own cloud - it's not possible as both Almanac structure and NAV3 messages are not documented and possible encrypted.

Indeed, and unfortunately to the best of my knowledge, the format of those messages has never been disclosed.

**Comment by StevenCellist** [comment]
@tve do you use this GNSS feature in any way?
